framework/tst/dslabs/framework/testing/runner/SimulatedNetwork.java

```diff
@@ -0,0 +1,202 @@
+package dslabs.framework.testing.runner;
+
+import dslabs.framework.Address;
+import dslabs.framework.testing.Event;
+import dslabs.framework.testing.MessageEnvelope;
+import dslabs.framework.testing.TimerEnvelope;
+import dslabs.framework.testing.runner.Network.Inbox;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.Random;
+import javax.annotation.Nonnull;
+import lombok.extern.java.Log;
+
+/**
+ * A deterministic simulated network.
+ */
+// the inheritance is for easily reusing `RunState`. consider extract common
+// behavior of `Network` and `SimulatedNetwork` into interface(s)
+@Log
+public class SimulatedNetwork extends Network {
+    @Override
+    @Nonnull
+    public Iterator<MessageEnvelope> iterator() {
+        throw new UnsupportedOperationException(); // TODO
+    }
+
+    static class VirtualTimeEvent implements Comparable<VirtualTimeEvent> {
+        long virtualTimeNanos;
+        Event inner;
+
+        @Override
+        public int compareTo(VirtualTimeEvent o) {
+            return Long.compare(virtualTimeNanos, o.virtualTimeNanos);
+        }
+    }
+
+    long virtualTimeNanos = 0;
+    long timeNanos = System.nanoTime();
+    boolean autoAdvance;
+    PriorityQueue<VirtualTimeEvent> events = new PriorityQueue<>();
+    HashMap<Address, Integer> numMessage = new HashMap<>();
+
+    Random rand = new Random();
+
+    @Override
+    public void send(MessageEnvelope messageEnvelope) {
+        var event = new VirtualTimeEvent();
+        event.virtualTimeNanos = (long) (virtualTimeNanos +
+                (38 * 1000 + 20 * 1000 * rand.nextGaussian()));
+        event.inner = new Event(messageEnvelope);
+        synchronized (this) {
+            events.add(event);
+            var n = numMessage.getOrDefault(messageEnvelope.to(), 0) + 1;
+            numMessage.put(messageEnvelope.to(), n);
+        }
+    }
+
+    void set(TimerEnvelope timerEnvelope) {
+        var event = new VirtualTimeEvent();
+        event.virtualTimeNanos = (long) (virtualTimeNanos +
+                (double) timerEnvelope.timerLengthMillis() * 1000 * 1000 +
+                (20 * 1000 * rand.nextGaussian()));
```

Comment:
This won't work as-is. The framework guarantees that timers with the same durations are delivered in the order they're set. (https://github.com/emichael/dslabs/tree/master/handout-files#search-tests)

Reply:
I agree. The consideration here is to make sure that if multiple nodes set timers with same length at the same time (for example when multiple clients send their first requests at virtual time 0), those timers should be delivered in any possible order. There should be per-node bookkeeping on timers to ensure the constraint you mentioned.